### PR TITLE
[change] Moved REST API topology collection DB write to background worker #163

### LINKED ISSUE #163: [change] Move REST API topology collection DB write logic to background worker
The `ReceiveTopologyView` is a bottleneck when instances start to grow because it takes time to process the data and respond.

For this reason we shall change the internal logic of the `Topology` model and `ReceiveTopologyView` so that in HTTP response cycle of `ReceiveTopologyView` we only perform the validation of the format of the incoming data, making sure `NetdiffException` is still raised if the format is invalid, but we shall move any logic that perform DB writes to new methods which are executed asynchronously via a celery worker.

We will need to add the dependency to celery on this module, let's do so by using openwisp-utils[celery]: https://github.com/openwisp/openwisp-utils/blob/master/setup.py#L67 (this centralizes the dependency definition of celery on a single module and it's necessary because we use celery in different modules and we want to make sure all the modules depend on the same version).

The celery worker must be launched using `transaction.on_commit` as we've been doing in other modules like OpenWISP Controller, eg: https://github.com/openwisp/openwisp-controller/search?q=transaction.on_commit .

### PR DESCRIPTION
**Pull request by Aryamanz29**
Closes #163

### CODE REVIEW

File: tests/openwisp2/settings.py [hunk 1]
```diff
@@ -164,9 +164,13 @@
 OPENWISP_ORGANIZATION_OWNER_ADMIN = True
 OPENWISP_USERS_AUTH_API = True
 
-CELERY_TASK_ALWAYS_EAGER = True
```
Comment (nemesisdesign) [inline tests/openwisp2/settings.py:167]: It's not great to do this here, we run all tests in eager mode across the repositories we maintain, why did you have to do this? Are you 100% is required? This is going to add overhead.

Comment (Aryamanz29) [reply tests/openwisp2/settings.py:167]: Yes, in tests we are still using [CELERY_TASK_ALWAYS_EAGER = True](https://github.com/openwisp/openwisp-network-topology/blob/0e23f1cccb284395e5541ac5b7f97a6d68762da7/tests/openwisp2/settings.py#L167-L172). I have used similar settings as those in other modules. Please see below:
 
https://github.com/openwisp/openwisp-controller/blob/a0c818a365c4dafcb1692653804d7d653e540540/tests/openwisp2/settings.py#L171-L176

https://github.com/openwisp/openwisp-monitoring/blob/25e00a23be37761622b1551419c63634b6c65f00/tests/openwisp2/settings.py#L168-L173

File: openwisp_network_topology/tests/test_topology.py [hunk 2]
```diff
@@ -451,13 +451,13 @@ def _assert_split_topology(self, topology):
...
             self.node_model.objects.all().delete()
-            t = self._set_receive(expiration_time=0.5)
+            t = self._set_receive(expiration_time=4)
             network1 = self._load('static/netjson-1-link.json')
             network2 = self._load('static/split-network.json')
             t.receive(network1)
             t.receive(network2)
             _assert_split_topology(self, t)
-            for sleep_time in [0.1, 0.25, 0.3]:
+            for sleep_time in [1, 2, 3]:
```
Comment (nemesisdesign) [inline openwisp_network_topology/tests/test_topology.py:460]: why is this huge increase in sleep time necessary?

File: tests/openwisp2/settings.py (hunk 1 above)
Comment (nemesisdesign) [reply tests/openwisp2/settings.py:167]: ok I read this too quckly, thanks!

File: tests/openwisp2/settings.py [hunk 3]
```diff
@@ -164,9 +164,13 @@
...
-CELERY_TASK_EAGER_PROPAGATES = True
-CELERY_BROKER_URL = 'memory://'
+if not TESTING:
+    CELERY_BROKER_URL = os.getenv('REDIS_URL', 'redis://localhost/1')
+else:
+    CELERY_TASK_ALWAYS_EAGER = True
+    CELERY_TASK_EAGER_PROPAGATES = True
+    CELERY_BROKER_URL = 'memory://'
```
Comment (nemesisdesign) [inline tests/openwisp2/settings.py:172]: Here's the problem with this change: it requires to upgrade the README instructions for users who want to run this app locally.
We could do that, but why even bother? I think we can just keep the eager mode, this also makes it easier for users to try this app. Maybe let's just add a comment above these settings like:

```
# the following 3 settings are for development only (don't use in production)
```

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
I have tested it and it's working as expected, nice!
Please see my comment below regarding celery eager mode.

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+version: "3"
+
+services:
+  redis:
+    image: redis:5.0-alpine
+    ports:
+      - "6379:6379"
+    entrypoint: redis-server --appendonly yes
```
Comment (nemesisdesign) [inline docker-compose.yml:8]: If this file is not mentioned in the README in any way, I think we must remove it.
We cannot keep files in the repositories which do not serve a specific purpose and are not explained anywhere.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

**Review by nemesisdesign — APPROVED** [review]

**Pull request merged**